### PR TITLE
[WIP] [Core] redirect stdout/stderr for all types of workers

### LINKED ISSUE #22664: [Feature] [Java] Redirect stdout and stderr to java-worker log file
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

Currently only lines printed via the slf4j logger will be written to the java-worker log file. Everything else through `System.out.println` or `System.err.println` still go through the file descriptor of raylet. It would great to redirect the log over to the worker where we have setup proper log streaming solution to the driver: https://github.com/ray-project/ray/pull/22629

cc @kfstorm 

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by kfstorm**
## Why are these changes needed?

Currently, only stdout/stderr of Python workers are redirected to log files. Java and C++ workers don't have any mechanism to redirect stdout/stderr to files. This PR tries to add redirection for Java and C++ workers by configuring redirection at the Raylet side (when starting new worker processes). The new mechanism is worker language ignorant.

TODO

- [ ] Windows support.
- [ ] Unify redirection for Python workers.
- [ ] Refine log file naming.
- [ ] Doc updates.

## Related issue number

Closes #22664

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by kfstorm** [comment]
@rkooo567 @simon-mo This PR is still in development. But you can comment on the high-level design.

**Comment by rkooo567** [comment]
OOC, Why is the current mechanism not working for java/cpp btw? I thought we were using Python subprocess to start those procs as well (and I believe the redirection is done by the Python subprocess APIs)?

**Comment by kfstorm** [comment]
> OOC, Why is the current mechanism not working for java/cpp btw? I thought we were using Python subprocess to start those procs as well (and I believe the redirection is done by the Python subprocess APIs)?

This is how we currently starts a worker process. https://github.com/ray-project/ray/blob/a70ada73411288a9c7a936dcc863872322486a3e/src/ray/raylet/worker_pool.cc#L578

It's purely C++ code. It uses `fork` and `exec...` system calls. See https://github.com/ray-project/ray/blob/a70ada73411288a9c7a936dcc863872322486a3e/src/ray/util/process.cc#L101

The redirection of Python worker logs happens in the `__main__` section of `default_worker.py`, which means that it only applies to Python workers. https://github.com/ray-project/ray/blob/a70ada73411288a9c7a936dcc863872322486a3e/python/ray/_private/workers/default_worker.py#L230-L234

**Comment by kfstorm** [comment]
@rkooo567 Any comments?

File: src/ray/raylet/worker_pool.cc [hunk 1]
```diff
@@ -434,10 +436,17 @@ std::tuple<Process, StartupToken> WorkerPool::StartWorkerProcess(
     // Make sure only the main thread is running in Python workers.
     env.insert({"RAY_start_python_importer_thread", "0"});
   }
-
+  std::string std_streams_redirect_file_prefix = log_dir_;
+  std_streams_redirect_file_prefix.append("/")
```
Comment (jjyao) [inline src/ray/raylet/worker_pool.cc:440]: "/" may not work for windows

File: src/ray/util/process.cc [hunk 2]
```diff
@@ -80,6 +80,24 @@ bool EnvironmentVariableLess::operator()(const std::string &a,
   return result;
 }
 
+void Redirect(const std::string &redirect_file, int id) {
+#ifdef _WIN32
```
Comment (jjyao) [inline src/ray/util/process.cc:84]: what about windows?

File: src/ray/util/process.cc [hunk 3]
```diff
@@ -80,6 +80,24 @@ bool EnvironmentVariableLess::operator()(const std::string &a,
...
+#else
+  mode_t pmode = S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH;
+  auto fdout = open(redirect_file.data(), O_RDWR | O_CREAT, pmode);
+  if (fdout < 0) {
+    RAY_LOG(WARNING) << "Create redirect file " << redirect_file << " failed.";
```
Comment (jjyao) [inline src/ray/util/process.cc:89]: We cannot continue if fdout < 0

**Review by jjyao — COMMENTED** [review]
LGTM at the high level.

Just make sure it works for Windows as well

**Pull request closed without merging**